code/part-two/client/source/App.jsx

```diff
@@ -1,22 +1,86 @@
 import React from 'react';
 import { Link, Route, Switch } from 'react-router-dom';
 
+import { SignupLogin } from './SignupLogin';
+
 import { Collection } from './Collection';
 import { CollectionList } from './CollectionList';
 import { Moji } from './Moji';
 import { Offer } from './Offer';
 import { OfferList } from './OfferList';
 
 export class App extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      privateKey: this.props.privateKey,
+      publicKey: this.props.publicKey
```

Comment:
I don't think `<App />` is being passed in any props related to `privateKey` and `publicKey` here, so maybe we should initialize them to `null` instead?